api/v1/mailables/MailableHandler.php

```diff
@@ -0,0 +1,153 @@
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class MailableHandler
+ * @ingroup api_v1_mailables
+ *
+ * @brief Base class to handle API requests for mailables.
+ */
+namespace PKP\api\v1\mailables;
+
+use APP\facades\Repo;
+use Illuminate\Support\Facades\Mail;
+use PKP\handler\APIHandler;
+use PKP\security\authorization\ContextRequiredPolicy;
+use PKP\security\authorization\PolicySet;
+use PKP\security\authorization\RoleBasedHandlerOperationPolicy;
+use PKP\security\Role;
+use Slim\Http\Request as SlimRequest;
+use Slim\Http\Response;
+
+class MailableHandler extends APIHandler
+{
+    /**
+     * @copydoc APIHandler::__construct()
+     */
+    public function __construct()
+    {
+        $this->_handlerPath = 'mailables';
+        $roles = [Role::ROLE_ID_SITE_ADMIN, Role::ROLE_ID_MANAGER];
+        $this->_endpoints = [
+            'GET' => [
+                [
+                    'pattern' => $this->getEndpointPattern(),
+                    'handler' => [$this, 'getMany'],
+                    'roles' => $roles,
+                ],
+                [
+                    'pattern' => $this->getEndpointPattern() . '/{className}',
+                    'handler' => [$this, 'get'],
+                    'roles' => $roles,
+                ],
+            ],
+            'PUT' => [
+                [
+                    'pattern' => $this->getEndpointPattern() . '/{className}/emailTemplates',
+                    'handler' => [$this, 'assignTemplates'],
+                    'roles' => $roles,
+                ],
+            ],
+        ];
+        parent::__construct();
+    }
+
+    /**
+     * @copydoc PKPHandler::authorize
+     */
+    public function authorize($request, &$args, $roleAssignments)
+    {
+        // This endpoint is not available at the site-wide level
+        $this->addPolicy(new ContextRequiredPolicy($request));
+
+        $rolePolicy = new PolicySet(PolicySet::COMBINING_PERMIT_OVERRIDES);
+        foreach ($roleAssignments as $role => $operations) {
+            $rolePolicy->addPolicy(new RoleBasedHandlerOperationPolicy($request, $role, $operations));
+        }
+        $this->addPolicy($rolePolicy);
+
+        return parent::authorize($request, $args, $roleAssignments);
+    }
+
+    /**
+     * Get an array of mailables
+     */
+    public function getMany(SlimRequest $slimRequest, Response $response, array $args): Response
+    {
+        $mailables = Repo::mailable()->getMany($this->getRequest()->getContext(), $slimRequest->getQueryParam('searchPhrase'));
+
+        return $response->withJson($mailables, 200);
+    }
+
+    /**
+     * Get a mailable by its class name
+     */
+    public function get(SlimRequest $slimRequest, Response $response, array $args): Response
+    {
+        $mailable = Repo::mailable()->getByClass($args['className']);
```

Comment:
Yes, the way how backslashes are escaped in JSON and URL is different. 

> Using class names directly could be a security risk. For example, as it currently stands, the API goes ahead and invokes $mailableClassName::getName() with any class name that I pass it

Yes, it shouldn't. I forgot to make the check prior to making any calls, same as in `assignTemplates`.

Reply:
I've rejected the idea with IDs because of the specifics of the Mailable entity. To work properly, it's required to create an entry in the database for each mailable and record the ID there to avoid clashes from Mailables added, e.g., by plugins.

Reply:
Ahh, I thought you meant simplify to a numeric ID.

Reply:
I've changed the logic so all Mailables with Configurable trait are having the `getId()` method